### PR TITLE
Add some cool structures

### PR DESCRIPTION
**Pull request by Lyxal**
- Actual if-statement
- Decision problem structure
- Generator structure

### CODE REVIEW

File: shared/src/main/scala/Globals.scala [hunk 1]
```diff
@@ -29,7 +29,7 @@ class Inputs(origInputs: Seq[VAny] = Seq.empty):
   private var origArr = origInputs.toArray.reverse
 
   /** Uses an array for constant access, not for mutating items */
-  private var currInputs = origArr
+  var currInputs = origArr
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/Globals.scala:32]: Well yes actually - it's called "I was doing some debugging and forgot to make it private again"

File: shared/src/main/scala/Interpreter.scala [hunk 2]
```diff
@@ -45,9 +45,33 @@ object Interpreter:
         elems.foreach(Interpreter.execute(_))
       case AST.CompositeNilad(elems) =>
         elems.foreach(Interpreter.execute(_))
-      case AST.If(thenBody, elseBody) =>
+      case AST.Ternary(thenBody, elseBody) =>
         if MiscHelpers.boolify(ctx.pop()) then execute(thenBody)
         else if elseBody.nonEmpty then execute(elseBody.get)
+
+      case AST.IfStatement(conds, bodies, None) =>
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/Interpreter.scala:52]: I think the version with None can be merged into the other one and just `.getorelse(some value)`

File: shared/src/main/scala/AST.scala [hunk 3]
```diff
@@ -18,12 +18,19 @@ enum AST(val arity: Option[Int]):
   case CompressedString(value: String) extends AST(Some(0))
   case CompressedNumber(value: String) extends AST(Some(0))
   case DictionaryString(value: String) extends AST(Some(0))
-  case If(thenBody: AST, elseBody: Option[AST]) extends AST(Some(1))
+  case Ternary(thenBody: AST, elseBody: Option[AST]) extends AST(Some(1))
+  case IfStatement(conds: List[AST], bodies: List[AST], elseBody: Option[AST])
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/AST.scala:22]: That may be too general - specific names are better in the AST definition.

The only real place `branches` makes sense is in lambdas

File: shared/src/main/scala/VList.scala [hunk 4]
```diff
@@ -126,4 +142,6 @@ object VList extends SpecificIterableFactory[VAny, VList]:
   override def fromSpecific(it: IterableOnce[VAny]): VList = new VList(
     it.iterator.toSeq
   )
+
+  def fromIterable(it: Seq[VAny]): VList = new VList(it)
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/VList.scala:146]: Sure. I didn't know what to name it so `from` sounds good

File: shared/src/main/scala/ListHelpers.scala [hunk 5]
```diff
@@ -27,6 +27,17 @@ object ListHelpers:
 
   end filter
 
+  def index(iterable: VList, ind: VAny)(using ctx: Context): VAny =
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/ListHelpers.scala:30]: What about the VList VList overload?

File: shared/src/main/scala/VList.scala [hunk 6]
```diff
@@ -53,6 +56,16 @@ class VList private (val lst: Seq[VAny])
...
       catch case _: IndexOutOfBoundsException => lst(ind % lst.length)
 
+  def applyBig(ind: BigInt): VAny =
+    if ind <= Int.MaxValue && ind >= Int.MinValue then return apply(ind.toInt)
+    val pos = VNum(ind.toString) % VNum(lst.length.toString)
+    var i = VNum("2147483647")
+    var temp = lst
+    while MiscHelpers.compare(i, pos) < 0 do
+      temp = temp.tail
+      i += VNum(1)
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/VList.scala:66]: That's just gonna overflow to negative numbers

[the remark this answers is not in the document]
Comment (mathcat4) [reply shared/src/main/scala/VList.scala:66]: I don't really understand the logic here, but wouldn't `val pos = VNum(ind) % VNum(lst.length)` error for infinite lists?

[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/VList.scala:66]: @mathcat4 ah dang didn't think of that.

[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/VList.scala:66]: the infinite indexing part. The `%` is for modular indexing.

[the remark this answers is not in the document]
Comment (mathcat4) [reply shared/src/main/scala/VList.scala]: oh, the index list-wrapping thing, right.

**Pull request merged**